tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1

```diff
@@ -66,38 +146,478 @@ function Get-AzProfile
   PROCESS
   {
     $Update = $PSBoundParameters.Update
-    # If Update is present, download ProfileMap from Storage blob endpoint
+    # If Update is present, download ProfileMap from online source
     if ($Update.IsPresent)
     {
       return (Get-AzureProfileMap)
     }
 
     # Check the cache
-    $ProfileCache = Get-ProfileCachePath
-    if(Test-Path $ProfileCache)
+    $LatestProfileMapPath = Get-LatestProfileMapPath
+    if(($LatestProfileMapPath -ne $null) -and (Test-Path $LatestProfileMapPath.FullName))
     {
-        $ProfileMap = Get-Content -Raw -Path "$ProfileCache\ProfileMap.json" -ErrorAction SilentlyContinue | ConvertFrom-Json 
-       
-        if ($ProfileMap -ne $null)
-        {
-           return $ProfileMap
-        }
+      $ProfileMap = Get-Content -Raw -Path $LatestProfileMapPath.FullName -ErrorAction SilentlyContinue | ConvertFrom-Json 
+      if ($ProfileMap -ne $null)
+      {
+         return $ProfileMap
+      }
     }
      
     # If cache doesn't exist, Check embedded source
     $defaults = [System.IO.Path]::GetDirectoryName($PSCommandPath)
     $ProfileMap = Get-Content -Raw -Path "$defaults\ProfileMap.json" -ErrorAction SilentlyContinue | ConvertFrom-Json
     if($ProfileMap -eq $null)
     {
-        # Cache & Embedded source empty; Return error and stop
-        throw [System.IO.FileNotFoundException] "$ProfileMap not found."
+      # Cache & Embedded source empty; Return error and stop
+      throw [System.IO.FileNotFoundException] "Profile meta data does not exist. Use 'Get-AzureRmProfile -Update' to download from online source."
     }
 
     return $ProfileMap
   }
 }
 
+# Lists the profiles available for install from gallery
+function Get-ProfilesAvailable
+{
+  param([parameter(Mandatory = $true)] [PSCustomObject] $ProfileMap)
+  $ProfileList = ""
+  foreach ($Profile in $ProfileMap) 
+  {
+    foreach ($Module in ($Profile | get-member -MemberType NoteProperty).Name)
+    {
+      $ProfileList += "Profile: $Module`n"
+      $ProfileList += "----------------`n"
+      $ProfileList += ($Profile.$Module | Format-List | Out-String)
+    } 
+  }
+  return $ProfileList
+}
+
+# Lists the profiles that are installed on the machine
+function Get-ProfilesInstalled
+{
+  param([parameter(Mandatory = $true)] [PSCustomObject] $ProfileMap, [REF]$IncompleteProfiles)
+  $result = @{}
+  $AllProfiles = ($ProfileMap | Get-Member -MemberType NoteProperty).Name
+
+  foreach ($key in $AllProfiles)
+  {
+    foreach ($module in ($ProfileMap.$key | Get-Member -MemberType NoteProperty).Name)
+    {
+      $versionList = $ProfileMap.$key.$module
+      foreach ($version in $versionList)
+      {
+        if ((Get-Module -Name $Module -ListAvailable | Where-Object { $_.Version -eq $version}) -ne $null)
```

Comment:
Get-Module without module name takes a long time (TotalSeconds      : 6.8812968) whereas with Module name parameter is much faster (TotalSeconds      : 0.1053396). So even if we have 20 modules, the total time taken would still be less than 2 seconds compared to "Get-Module -ListAvailable" which is 6.88 seconds.